network/src/peer_manager/mod.rs

```diff
@@ -202,6 +184,31 @@ impl PeerManagerRequestSender {
     }
 }
 
+impl ConnectionRequestSender {
+    /// Construct a new ConnectionRequestSender with a raw libra_channel::Sender
+    pub fn new(inner: libra_channel::Sender<PeerId, ConnectionRequest>) -> Self {
+        Self { inner }
+    }
+
+    pub async fn dial_peer(
+        &mut self,
+        peer: PeerId,
+        addr: Multiaddr,
+    ) -> Result<(), PeerManagerError> {
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        self.inner
+            .push(peer, ConnectionRequest::DialPeer(peer, addr, oneshot_tx))?;
+        oneshot_rx.await?
+    }
+
+    pub async fn disconnect_peer(&mut self, peer: PeerId) -> Result<(), PeerManagerError> {
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        self.inner
+            .push(peer, ConnectionRequest::DisconnectPeer(peer, oneshot_tx))?;
+        oneshot_rx.await?
+    }
```

Comment:
random thought: something i've noticed with this style API (same as `send_rpc`, prev. `unary_rpc`) is that it basically forces a sender clone every time if you ever want to send more than one request at a time. i've been wondering how we might change the API so that the send req/await resp happen in two separate steps. in other words, without cloning, you effectively "lock" your sender with the &mut ref for the duration of the entire rpc request/response instead of just pushing the request (not async) and later awaiting the response (async).

~~of course, the annoying part about splitting the send/await part is you have to name the Future you want to return, which quickly gets unwieldy, unless we `Pin<Box<dyn Future ...>` everything.~~

nvm, since our senders are not generic, I think we can actually return `impl Future`.

for example, splitting `disconnect_peer` might look like:

```rust
impl ConnectionRequestSender {
    pub fn disconnect_peer(&mut self, peer: PeerId) -> impl Future<Output = Result<(), PeerManagerError>> {
        let (oneshot_tx, oneshot_rx) = oneshot::channel();
        self.inner
            .push(peer, ConnectionRequest::DisconnectPeer(peer, oneshot_tx))?;
        oneshot_rx.map_err(Into::into)
    }
}
```

then, if the client is just doing one-at-a-time, they can still do exactly what they're doing today

```rust
let mut conn_req_tx: ConnectionRequestSender = /* .. */;

conn_req_tx.disconnect_peer(peer_id).await?;
```

but if they need multiple requests concurrently, they can separate the send req/await resp steps

Reply:
if we want to do this refactoring, we should probably do it in a separate diff : )

Reply:
yeah, great observation. Surprised that this requirement has never come up :-o. I was also discussing with @davidiw about making the `NetworkSender` generic to avoid the code duplication in all the clients. We might still be able to get by using the `async_trait` macro.

Reply:
Interesting!